Content.Server/Body/Systems/RespiratorSystem.cs

```diff
@@ -41,211 +56,140 @@ public override void Update(float frameTime)
                     continue;
                 }
 
                 StopSuffocation(uid, respirator);
+                respirator.SuffocationCycles = 0;
             }
         }
 
-        private Dictionary<Gas, float> NeedsAndDeficit(RespiratorComponent respirator)
-        {
-            var needs = new Dictionary<Gas, float>(respirator.NeedsGases);
-            foreach (var (gas, amount) in respirator.DeficitGases)
-            {
-                var newAmount = (needs.GetValueOrDefault(gas) + amount);
-                needs[gas] = newAmount;
-            }
-
-            return needs;
-        }
-
-        private void ClampDeficit(RespiratorComponent respirator)
+        public void Inhale(EntityUid uid, SharedBodyComponent? body=null)
         {
-            var deficitGases = new Dictionary<Gas, float>(respirator.DeficitGases);
-
-            foreach (var (gas, deficit) in deficitGases)
-            {
-                if (!respirator.NeedsGases.TryGetValue(gas, out var need))
-                {
-                    respirator.DeficitGases.Remove(gas);
-                    continue;
-                }
+            if (!Resolve(uid, ref body, false))
+                return;
 
-                if (deficit > need)
-                {
-                    respirator.DeficitGases[gas] = need;
-                }
-            }
-        }
+            var organs = _bodySystem.GetComponentsOnMechanisms<LungComponent>(uid, body).ToArray();
 
-        private float SuffocatingPercentage(RespiratorComponent respirator)
-        {
-            var total = 0f;
+            // Inhale gas
+            var ev = new InhaleLocationEvent();
+            RaiseLocalEvent(uid, ev, false);
 
-            foreach (var (gas, deficit) in respirator.DeficitGases)
+            if (ev.Gas == null)
             {
-                var lack = 1f;
-                if (respirator.NeedsGases.TryGetValue(gas, out var needed))
-                {
-                    lack = deficit / needed;
-                }
-
-                total += lack / Atmospherics.TotalNumberOfGases;
+                ev.Gas = _atmosSys.GetTileMixture(Transform(uid).Coordinates);
+                if (ev.Gas == null) return;
             }
 
-            return total;
-        }
-
-        private float GasProducedMultiplier(RespiratorComponent respirator, Gas gas, float usedAverage)
-        {
-            if (!respirator.ProducesGases.TryGetValue(gas, out var produces))
-            {
-                return 0;
-            }
+            var ratio = (Atmospherics.BreathVolume / ev.Gas.Volume);
+            var actualGas = ev.Gas.RemoveRatio(ratio);
 
-            if (!respirator.NeedsGases.TryGetValue(gas, out var needs))
+            var lungRatio = 1.0f / organs.Length;
+            foreach (var (lung, _) in organs)
             {
-                needs = 1;
+                var gas = organs.Length == 1 ? actualGas : actualGas.RemoveRatio(lungRatio);
```

Comment:
Given that the `.Merge` doesn't remove air from the donor, this could just be moved outside of the for loop, right? 

Also, I think at the moment if you had two lungs, the first lung gets half of `actualGas`, while the second one gets half of what is left (i.e., 1/4)?

Reply:
oh good point